File: android/app/src/main/java/com/appxstudios/festivalconnection/services/WalletManager.kt
package com.appxstudios.festivalconnection.services

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import breez_sdk_liquid.*
import com.appxstudios.festivalconnection.BuildConfig
import com.appxstudios.festivalconnection.models.WalletTransaction
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import java.io.File
import java.security.SecureRandom
import java.util.UUID

object WalletManager : EventListener {
    private var sdk: BindingLiquidSdk? = null
    private var eventListenerId: String? = null
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    private val _balanceSat = MutableStateFlow(0L)
    val balanceSat: StateFlow<Long> = _balanceSat

    private val _balanceUSD = MutableStateFlow(0.0)
    val balanceUSD: StateFlow<Double> = _balanceUSD

    private val _transactions = MutableStateFlow<List<WalletTransaction>>(emptyList())
    val transactions: StateFlow<List<WalletTransaction>> = _transactions

    private val _isConnected = MutableStateFlow(false)
    val isConnected: StateFlow<Boolean> = _isConnected

    private val _connectionError = MutableStateFlow<String?>(null)
    val connectionError: StateFlow<String?> = _connectionError

    fun connect(context: Context) {
        if (sdk != null) return
        scope.launch {
            try {
                val mnemonic = getOrCreateMnemonic(context)
                val config = defaultConfig(LiquidNetwork.MAINNET, BuildConfig.BREEZ_API_KEY)
                config.workingDir = walletDirectory(context)
                val req = ConnectRequest(config, mnemonic)
                sdk = breez_sdk_liquid.connect(req)
                eventListenerId = sdk?.addEventListener(this@WalletManager)
                _isConnected.value = true
                _connectionError.value = null
                refreshBalance()
                refreshTransactions()
            } catch (e: Exception) {
                _connectionError.value = e.message
                _isConnected.value = false
            }
        }
    }

    fun disconnect() {
        try {
            eventListenerId?.let { sdk?.removeEventListener(it) }
            eventListenerId = null
            sdk?.disconnect()
        } catch (_: Exception) {}
        sdk = null
        _isConnected.value = false
    }

    // MARK: - Balance

    fun refreshBalance() {
        scope.launch {
            try {
                val info = sdk?.getInfo() ?: return@launch
                _balanceSat.value = info.walletInfo.balanceSat.toLong()
                try {
                    val rates = sdk?.fetchFiatRates()
                    val usdRate = rates?.firstOrNull { it.coin == "USD" }
                    if (usdRate != null) {
                        _balanceUSD.value = _balanceSat.value / 100_000_000.0 * usdRate.value
                    }
                } catch (_: Exception) {}
            } catch (e: Exception) {
                println("[Wallet] Balance refresh failed: ${e.message}")
            }
        }
    }

    // MARK: - Receive (Create Invoice)

    suspend fun createInvoice(amountSat: Long, description: String = ""): String {
        val sdk = sdk ?: throw WalletException("Wallet not connected")
        val prepareReq = PrepareReceiveRequest(
            paymentMethod = PaymentMethod.BOLT11_INVOICE,
            amount = ReceiveAmount.Bitcoin(payerAmountSat = amountSat.toULong())
        )
        val prepareResponse = sdk.prepareReceivePayment(prepareReq)
        val receiveReq = ReceivePaymentRequest(
            prepareResponse = prepareResponse,
            description = description
        )
        val response = sdk.receivePayment(receiveReq)
        return response.destination
    }

    // MARK: - Send (Pay Invoice)

    suspend fun sendPayment(invoice: String, amountSat: Long? = null) {
        val sdk = sdk ?: throw WalletException("Wallet not connected")
        val prepareReq = if (amountSat != null) {
            PrepareSendRequest(destination = invoice, amount = PayAmount.Bitcoin(receiverAmountSat = amountSat.toULong()))
        } else {
            PrepareSendRequest(destination = invoice)
        }
        val prepareResponse = sdk.prepareSendPayment(prepareReq)
        val sendReq = SendPaymentRequest(prepareResponse = prepareResponse)
        sdk.sendPayment(sendReq)
        refreshBalance()
        refreshTransactions()
    }

    // MARK: - Parse Invoice

    fun parseInvoice(input: String): Pair<Long?, String?> {
        val sdk = sdk ?: throw WalletException("Wallet not connected")
        return when (val parsed = sdk.parse(input)) {
            is InputType.Bolt11 -> {
                val amountSat = parsed.invoice.amountMsat?.let { it.toLong() / 1000 }
                Pair(amountSat, parsed.invoice.description)
            }
            else -> Pair(null, null)
        }
    }

    // MARK: - Transaction History

    fun refreshTransactions() {
        scope.launch {
            try {
                val payments = sdk?.listPayments(ListPaymentsRequest()) ?: return@launch
                _transactions.value = payments.mapNotNull { payment ->
                    val direction = when (payment.paymentType) {
                        PaymentType.RECEIVE -> "received"
                        else -> "sent"
                    }
                    val desc = when (val details = payment.details) {
                        is PaymentDetails.Lightning -> details.description
                        is PaymentDetails.Liquid -> details.description
                        is PaymentDetails.Bitcoin -> details.description
                        else -> ""
                    }
                    WalletTransaction(
                        id = payment.txId ?: UUID.randomUUID().toString(),
                        amountSat = payment.amountSat.toLong(),
                        amountUSD = 0.0,
                        direction = direction,
                        timestamp = (payment.timestamp?.toLong() ?: 0L) * 1000L,
                        description = desc,
                        paymentHash = payment.txId ?: "",
                        status = payment.status.toString(),
                        fees = payment.feesSat.toLong()
                    )
                }
                // Apply fiat rates
                try {
                    val rates = sdk?.fetchFiatRates()
                    val usdRate = rates?.firstOrNull { it.coin == "USD" }
                    if (usdRate != null) {
                        _transactions.value = _transactions.value.map { tx ->
                            tx.copy(amountUSD = tx.amountSat.toDouble() / 100_000_000.0 * usdRate.value)
                        }
                    }
                } catch (_: Exception) {}
            } catch (e: Exception) {
                println("[Wallet] Transaction refresh failed: ${e.message}")
            }
        }
    }

    // MARK: - Event Listener

    override fun onEvent(e: SdkEvent) {
        when (e) {
            is SdkEvent.PaymentSucceeded, is SdkEvent.PaymentPending -> {
                refreshBalance()
                refreshTransactions()
            }
            is SdkEvent.Synced -> refreshBalance()
            else -> {}
        }
    }

    // MARK: - Mnemonic Management

    private fun getOrCreateMnemonic(context: Context): String {
        val prefs = getEncryptedPrefs(context)
        val existing = prefs.getString("wallet_mnemonic", null)
        // Guard: old versions stored raw hex (64 chars, no spaces). Treat those as invalid and regenerate.
        if (existing != null && existing.trim().split(" ").size in 12..24) {
            return existing
        }

        // Generate real BIP-39 12-word mnemonic (128 bits of entropy)
        // Uses https://github.com/Electric-Coin-Company/kotlin-bip39 — accepted by Breez SDK Liquid
        val code = cash.z.ecc.android.bip39.Mnemonics.MnemonicCode(
            cash.z.ecc.android.bip39.Mnemonics.WordCount.COUNT_12
        )
        val mnemonic = String(code.chars)
        prefs.edit().putString("wallet_mnemonic", mnemonic).apply()
        return mnemonic
    }

    private fun getEncryptedPrefs(context: Context): SharedPreferences {
        val masterKey = MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
        return EncryptedSharedPreferences.create(
            context, "fc_wallet_prefs", masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        )
    }

    private fun walletDirectory(context: Context): String {
        val dir = File(context.filesDir, "breez-wallet")
        if (!dir.exists()) dir.mkdirs()
        return dir.absolutePath
    }

    class WalletException(message: String) : Exception(message)
}

File: ios/FestivalConnection/Services/WalletManager.swift
import Foundation
import BreezSDKLiquid

@MainActor
final class WalletManager: ObservableObject {
    static let shared = WalletManager()

    @Published var balanceSat: UInt64 = 0
    @Published var balanceUSD: Double = 0.0
    @Published var transactions: [WalletTransaction] = []
    @Published var isConnected: Bool = false
    @Published var connectionError: String?
    /// Live sats-per-USD rate, refreshed alongside the balance from Breez fetchFiatRates().
    /// Falls back to ~1100 sats/$ until the first successful fetch.
    @Published var satPerUSD: Double = 1100

    private var sdk: BindingLiquidSdk?
    private var eventListenerId: String?
    private let keychainAccount = "fc_wallet_mnemonic"

    // MARK: - Lifecycle

    func connect() {
        guard sdk == nil else { return }
        do {
            let mnemonic = getOrCreateMnemonic()
            var config = try defaultConfig(network: LiquidNetwork.mainnet, breezApiKey: "MIIBeTCCASugAwIBAgIHPxUDqQ19mTAFBgMrZXAwEDEOMAwGA1UEAxMFQnJlZXowHhcNMjYwNDA3MjMwMTM5WhcNMzYwNDA0MjMwMTM5WjAvMRQwEgYDVQQKEwtBUFBYU3R1ZGlvczEXMBUGA1UEAxMOQWJyYWhhbSBWYXJnYXMwKjAFBgMrZXADIQDQg")
            config.workingDir = walletDirectory()
            let req = ConnectRequest(config: config, mnemonic: mnemonic)
            sdk = try BreezSDKLiquid.connect(req: req)
            isConnected = true
            connectionError = nil
            // Set up event listener — addEventListener returns an ID string
            eventListenerId = try sdk?.addEventListener(listener: self)
            // Initial data fetch
            refreshBalance()
            refreshTransactions()
        } catch {
            connectionError = error.localizedDescription
            isConnected = false
        }
    }

    func disconnect() {
        if let listenerId = eventListenerId {
            try? sdk?.removeEventListener(id: listenerId)
            eventListenerId = nil
        }
        try? sdk?.disconnect()
        sdk = nil
        isConnected = false
    }

    // MARK: - Balance

    func refreshBalance() {
        guard let sdk = sdk else { return }
        Task {
            do {
                let info = try sdk.getInfo()
                // walletInfo.balanceSat is UInt64 (non-optional)
                let btcBalance = info.walletInfo.balanceSat
                balanceSat = btcBalance
                // Fetch fiat rate for USD conversion
                if let rates = try? sdk.fetchFiatRates() {
                    if let usdRate = rates.first(where: { $0.coin == "USD" }) {
                        balanceUSD = Double(btcBalance) / 100_000_000.0 * usdRate.value
                        // usdRate.value = USD per BTC. 1 BTC = 100_000_000 sats.
                        // sats per $1 = 100_000_000 / USD-per-BTC
                        if usdRate.value > 0 {
                            satPerUSD = 100_000_000.0 / usdRate.value
                        }
                    }
                }
            } catch {
                print("[Wallet] Balance refresh failed: \(error)")
            }
        }
    }

    // MARK: - Receive (Create Invoice)

    func createInvoice(amountSat: UInt64, description: String = "") async throws -> String {
        guard let sdk = sdk else { throw WalletError.notConnected }
        // PrepareReceiveRequest takes paymentMethod + optional ReceiveAmount
        // ReceiveAmount.bitcoin uses payerAmountSat:
        let prepareReq = PrepareReceiveRequest(
            paymentMethod: .bolt11Invoice,
            amount: .bitcoin(payerAmountSat: amountSat)
        )
        let prepareResponse = try sdk.prepareReceivePayment(req: prepareReq)
        // ReceivePaymentRequest takes prepareResponse (not individual fields)
        let receiveReq = ReceivePaymentRequest(
            prepareResponse: prepareResponse,
            description: description
        )
        let response = try sdk.receivePayment(req: receiveReq)
        return response.destination
    }

    // MARK: - Send (Pay Invoice)

    func sendPayment(invoice: String, amountSat: UInt64? = nil) async throws {
        guard let sdk = sdk else { throw WalletError.notConnected }
        // PrepareSendRequest takes destination + optional PayAmount
        // PayAmount.bitcoin uses receiverAmountSat:
        let prepareReq: PrepareSendRequest
        if let amt = amountSat {
            prepareReq = PrepareSendRequest(destination: invoice, amount: .bitcoin(receiverAmountSat: amt))
        } else {
            prepareReq = PrepareSendRequest(destination: invoice)
        }
        let prepareResponse = try sdk.prepareSendPayment(req: prepareReq)
        // SendPaymentRequest takes prepareResponse (not destination)
        let sendReq = SendPaymentRequest(prepareResponse: prepareResponse)
        let _ = try sdk.sendPayment(req: sendReq)
        // Refresh after payment
        refreshBalance()
        refreshTransactions()
    }

    // MARK: - Parse Invoice

    func parseInvoice(_ input: String) throws -> (amountSat: UInt64?, description: String?) {
        guard let sdk = sdk else { throw WalletError.notConnected }
        let parsed = try sdk.parse(input: input)
        switch parsed {
        case .bolt11(let invoice):
            // LnInvoice has amountMsat: UInt64? and description: String?
            return (amountSat: invoice.amountMsat.map { $0 / 1000 }, description: invoice.description)
        case .bolt12Offer(let offer, _):
            // bolt12Offer has two associated values: (offer: LnOffer, bip353Address: String?)
            // LnOffer.minAmount is Amount? with case .bitcoin(amountMsat:)
            var offerAmountSat: UInt64? = nil
            if let minAmount = offer.minAmount {
                switch minAmount {
                case .bitcoin(let amountMsat):
                    offerAmountSat = amountMsat / 1000
                case .currency:
                    break
                }
            }
            return (amountSat: offerAmountSat, description: offer.description)
        default:
            return (amountSat: nil, description: nil)
        }
    }

    // MARK: - Transaction History

    func refreshTransactions() {
        guard let sdk = sdk else { return }
        Task {
            do {
                let payments = try sdk.listPayments(req: ListPaymentsRequest())
                transactions = payments.compactMap { payment -> WalletTransaction? in
                    let direction: String
                    switch payment.paymentType {
                    case .receive: direction = "received"
                    case .send: direction = "sent"
                    }
                    // Extract description from PaymentDetails (Payment has no .description)
                    let desc: String
                    switch payment.details {
                    case .lightning(_, let description, _, _, _, _, _, _, _, _, _, _, _, _, _):
                        desc = description
                    case .liquid(_, _, let description, _, _, _, _):
                        desc = description
                    case .bitcoin(_, _, let description, _, _, _, _, _, _, _):
                        desc = description
                    }
                    return WalletTransaction(
                        id: payment.txId ?? UUID().uuidString,
                        amountSat: payment.amountSat,
                        amountUSD: 0,  // Will be calculated with fiat rate
                        direction: direction,
                        timestamp: Date(timeIntervalSince1970: TimeInterval(payment.timestamp)),
                        description: desc,
                        paymentHash: payment.txId ?? "",
                        status: String(describing: payment.status),
                        fees: payment.feesSat
                    )
                }
                // Apply fiat rates to transactions
                if let rates = try? sdk.fetchFiatRates() {
                    if let usdRate = rates.first(where: { $0.coin == "USD" }) {
                        for i in self.transactions.indices {
                            self.transactions[i].amountUSD = Double(self.transactions[i].amountSat) / 100_000_000.0 * usdRate.value
                        }
                    }
                }
            } catch {
                print("[Wallet] Transaction refresh failed: \(error)")
            }
        }
    }

    // MARK: - Mnemonic Management

    /// Returns a BIP-39 mnemonic string, generating + persisting one on first run.
    /// Old raw-byte "seeds" stored under this same Keychain account are intentionally
    /// not migrated — they were never valid BIP-39 phrases, so any funds attached
    /// would have been unrecoverable cross-device anyway.
    private func getOrCreateMnemonic() -> String {
        if let existing = loadMnemonicFromKeychain(), BIP39.isValidMnemonic(existing) {
            return existing
        }
        let mnemonic = BIP39.generate12WordMnemonic()
        saveMnemonicToKeychain(mnemonic)
        return mnemonic
    }

    private func saveMnemonicToKeychain(_ mnemonic: String) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: keychainAccount,
            kSecAttrService as String: "com.appxstudios.festivalconnection.wallet",
            kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlock,
        ]
        SecItemDelete(query as CFDictionary)
        var addQuery = query
        addQuery[kSecValueData as String] = Data(mnemonic.utf8)
        SecItemAdd(addQuery as CFDictionary, nil)
    }

    private func loadMnemonicFromKeychain() -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: keychainAccount,
            kSecAttrService as String: "com.appxstudios.festivalconnection.wallet",
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        guard status == errSecSuccess, let data = result as? Data,
              let str = String(data: data, encoding: .utf8) else { return nil }
        return str
    }

    private func walletDirectory() -> String {
        let dir = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
            .appendingPathComponent("breez-wallet", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir.path
    }

    enum WalletError: Error, LocalizedError {
        case notConnected
        var errorDescription: String? {
            switch self {
            case .notConnected: return "Wallet not connected"
            }
        }
    }
}

// MARK: - Event Listener
extension WalletManager: EventListener {
    nonisolated func onEvent(e: SdkEvent) {
        Task { @MainActor in
            switch e {
            case .paymentSucceeded(_), .paymentPending(_):
                refreshBalance()
                refreshTransactions()
            case .synced:
                refreshBalance()
            default:
                break
            }
        }
    }
}
